pkg/csi/service/vanilla/controller.go

```diff
@@ -383,75 +383,178 @@ func (c *controller) ReloadConfiguration() error {
 	if err != nil {
 		return logger.LogNewErrorf(log, "failed to read config. Error: %+v", err)
 	}
-	newVCConfig, err := cnsvsphere.GetVirtualCenterConfig(ctx, cfg)
-	if err != nil {
-		log.Errorf("failed to get VirtualCenterConfig. err=%v", err)
-		return err
-	}
-	if newVCConfig != nil {
-		var vcenter *cnsvsphere.VirtualCenter
-		if c.manager.VcenterConfig.Host != newVCConfig.Host ||
-			c.manager.VcenterConfig.Username != newVCConfig.Username ||
-			c.manager.VcenterConfig.Password != newVCConfig.Password {
-
-			// Verify if new configuration has valid credentials by connecting to
-			// vCenter. Proceed only if the connection succeeds, else return error.
-			newVC := &cnsvsphere.VirtualCenter{Config: newVCConfig}
-			if err = newVC.Connect(ctx); err != nil {
-				return logger.LogNewErrorf(log, "failed to connect to VirtualCenter host: %q, Err: %+v",
-					newVCConfig.Host, err)
+	if multivCenterCSITopologyEnabled {
+		var multivCenterTopologyDeployment bool
+		if len(c.managers.VcenterConfigs) > 1 {
+			multivCenterTopologyDeployment = true
+		}
+		newVcenterConfigs, err := cnsvsphere.GetVirtualCenterConfigs(ctx, cfg)
+		if err != nil {
+			return logger.LogNewErrorf(log, "failed to get VirtualCenterConfigs. err=%v", err)
+		}
+		if newVcenterConfigs != nil {
+			for _, newVCConfig := range newVcenterConfigs {
+				var vcenter *cnsvsphere.VirtualCenter
+				oldvcconfig, found := c.managers.VcenterConfigs[newVCConfig.Host]
+				if !found || oldvcconfig.Username != newVCConfig.Username || oldvcconfig.Password != newVCConfig.Password {
+					newVC := &cnsvsphere.VirtualCenter{Config: newVCConfig}
+					if err = newVC.Connect(ctx); err != nil {
+						return logger.LogNewErrorf(log, "failed to connect to VirtualCenter host: %q, Err: %+v",
+							newVCConfig.Host, err)
+					}
+					// Reset vCenter singleton instance by passing reload flag as true.
+					log.Info("Obtaining new vCenterInstance using new credentials")
+					vcenter, err = cnsvsphere.GetVirtualCenterInstanceForVCenterConfig(ctx, newVCConfig, true)
+					if err != nil {
+						return logger.LogNewErrorf(log, "failed to get vCenterInstance for vCenter Host: "+
+							"%q, err: %v", newVCConfig.Host, err)
+					}
+				} else {
+					// If it's not VC credentials update, same singleton
+					// instance can be used and it's Config field can be updated.
+					vcenter, err = cnsvsphere.GetVirtualCenterInstance(ctx, &cnsconfig.ConfigurationInfo{Cfg: cfg}, false)
+					if err != nil {
+						return logger.LogNewErrorf(log, "failed to get VirtualCenter. err=%v", err)
+					}
+					vcenter.Config = newVCConfig
```

Comment:
Do you think calling a connect here will be helpful?

Reply:
no, we don't need to connect here, as we are just refreshing other config fields here.